File: 07_LinkedList/python/linked_list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def append(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            return
        last = self.head
        while last.next:
            last = last.next
        last.next = new_node

    def get_at(self, index):
        current = self.head
        count = 0
        while current:
            if count == index:
                return current.data
            count += 1
            current = current.next
        return None

if __name__ == "__main__":
    ll = LinkedList()
    ll.append(10)
    ll.append(20)
    
    assert ll.get_at(0) == 10, "Erreur à l'index 0"
    assert ll.get_at(1) == 20, "Erreur à l'index 1"
    print("Test Python Linked List réussi !")

File: 07_LinkedList/c/linked_list.c
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

typedef struct Node {
    int data;
    struct Node* next;
} Node;

void append(Node** head_ref, int new_data) {
    Node* new_node = (Node*)malloc(sizeof(Node));
    Node* last = *head_ref;
    new_node->data = new_data;
    new_node->next = NULL;

    if (*head_ref == NULL) {
        *head_ref = new_node;
        return;
    }

    while (last->next != NULL) {
        last = last->next;
    }
    last->next = new_node;
}

int getAt(Node* head, int index) {
    Node* current = head;
    int count = 0;
    while (current != NULL) {
        if (count == index) return current->data;
        count++;
        current = current->next;
    }
    return -1; 
}

void freeList(Node* head) {
    Node* temp;
    while (head != NULL) {
        temp = head;
        head = head->next;
        free(temp);
    }
}

int main() {
    Node* head = NULL;
    append(&head, 10);
    append(&head, 20);

    assert(getAt(head, 0) == 10);
    assert(getAt(head, 1) == 20);

    printf("Test C Linked List réussi !\n");
    freeList(head);
    return 0;
}

